File: KunalSheet/src/main/java/org/ds/array/medium/NextPermutation.java
package org.ds.array.medium;

import java.util.Arrays;

public class NextPermutation {
    public static void main(String[] args){
        int[] nums = {1,2,3}; // o/p: [1,3,2]
        int[] num1 = {3,2,1}; // o/p: [1,2,3]
        nextPermutation(nums);
        nextPermutation(num1);
        System.out.println(Arrays.toString(nums));
        System.out.println(Arrays.toString(num1));
    }

    public static void nextPermutation(int[] nums) {
         int i = nums.length - 2;
        while (i >= 0 && nums[i + 1] <= nums[i]) {
            i--;
        }
        if (i >= 0) {
            int j = nums.length - 1;
            while (nums[j] <= nums[i]) {
                j--;
            }
            swap(nums, i, j);
        }
        reverse(nums, i + 1);
    }

    private static void swap(int[] arr, int i, int j){
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    private  static void reverse(int[] nums, int start) {
        int i = start, j = nums.length - 1;
        while (i < j) {
            swap(nums, i, j);
            i++;
            j--;
        }
    }
}

File: Python-version/Medium/Array/next_permutation.py
from typing import List

class NextPermutation:
    def next_permutation(self, arr:List[int])-> None:
        i = len(arr) - 2
        while i >=0 and arr[i] >= arr[i+1]:
            i -= 1
        
        if i >=0 :
            j = len(arr) - 1
            while arr[j] > arr[i]:
                j -= 1
            self._swap(arr,i,j)
        
        self._reverse(arr,i+1)
        
    
    def _swap(self, arr:List[int], i:int, j:int)-> None:
        arr[i], arr[j] = arr[j], arr[i]

    def _reverse(self, arr:List[int], start:int)-> None:
        i , j = start, len(arr) - 1
        while i<j:
            self._swap(arr,i,j)
            i += 1
            j -= 1

def main():
    solution = NextPermutation()

    input = [3,2,1]
    solution.next_permutation(input)
    print(input)

if __name__ == "__main__":
    main()